components/raftstore/src/store/fsm/store.rs

```diff
@@ -1500,14 +1502,19 @@ impl<'a, T: Transport, C: PdClient> StoreFsmDelegate<'a, T, C> {
             if util::is_first_vote_msg(msg.get_message()) {
                 meta.pending_votes.push(msg.to_owned());
             }
-
-            if maybe_destroy_source(
+            let (can_destroy, merge_to_this_peer) = maybe_destroy_source(
                 meta,
                 region_id,
+                msg.get_to_peer().get_id(),
                 exist_region.get_id(),
                 msg.get_region_epoch().to_owned(),
-            ) {
-                regions_to_destroy.push(exist_region.get_id());
+            );
+            if can_destroy {
+                if !merge_to_this_peer {
+                    regions_to_destroy.push(exist_region.get_id());
+                } else {
+                    // TODO, it shouldn't happen because all target peer should exist during merging
```

Comment:
We should print some error log in this case.

Reply:
Done.